pre-check/v1.0.3/check.sh

```diff
@@ -0,0 +1,247 @@
+    modifiedStatus:
+      - apiVersion: v1
+        kind: ConfigMap
+        missing: true
+        name: longhorn-storageclass
+        namespace: longhorn-system
+    name: fleet-local/local
+ready: 0
+EOF
+
+    if ! diff <(yq -P 'sort_keys(..)' $current_summary) <(yq -P 'sort_keys(..)' $expected_summary); then
+        echo "Harvester bundle is not ready!"
+        cat $current_summary
+        record_fail
+        return
+    fi
+
+    echo "The Harvester bundle is ready."
+}
+
+check_nodes()
+{
+    local failed="false"
+    echo ">>> Check all nodes are ready..."
+
+    # nodes should not be cordoned
+    nodes=$(kubectl get nodes -o yaml)
+    unschedulable=$(echo "$nodes" | yq '.items | any_c(.spec.unschedulable == true)')
+
+    if [ "$unschedulable" = "true" ]; then
+        echo "There are unschedulable nodes:"
+        echo "$nodes" | yq '.items[] | select(.spec.unschedulable == true)  | .metadata.name'
+        failed="true"
+    fi
+
+    # nodes should be ready
+    echo "$nodes" | yq .items[].metadata.name |
+        while read -r node_name; do
+            node_ready=$(kubectl get nodes $node_name -o yaml | yq '.status.conditions | any_c(.type == "Ready" and .status == "True")')
+
+            if [ "$node_ready" = "false" ]; then
+                echo "Node $node_name is not ready!"
+                failed="true"
+            fi
+        done
+
+    if [ "$failed" = "true" ]; then
+        record_fail
+    else
+        echo "All nodes are ready."
+    fi
+}
+
+check_cluster()
+{
+    echo ">>> Check the CAPI cluster is provisioned..."
+    cluster_phase=$(kubectl get clusters.cluster.x-k8s.io/local -n fleet-local -o yaml | yq '.status.phase')
+
+    # cluster should be in provisioned
+    if [ "$cluster_phase" != "Provisioned" ]; then
+        echo "Cluster is not provisioned ($cluster_phase)"
+        record_fail
+        return
+    fi
+
+    echo "The CAPI cluster is provisioned."
+}
+
+check_machines()
+{
+    local failed="false"
+    echo ">>> Check the CAPI machines are running..."
+
+    # all machines need to be provisioned
+    kubectl get machines.cluster.x-k8s.io -n fleet-local -o yaml | yq '.items[].metadata.name' |
+        while read -r machine_name; do
+            machine_phase=$(kubectl get machines.cluster.x-k8s.io/$machine_name -n fleet-local -o yaml | yq '.status.phase')
+
+            if [ "$machine_phase" != "Running" ]; then
+                echo "CAPI machine $machine_name phase is not Running ($machine_phase)."
+                failed="true"
+            fi
+        done
+
+    if [ "$failed" = "true" ]; then
+        record_fail
+    else
+        echo "The CAPI machines are provisioned."
+    fi
+}
+
+check_volumes()
+{
+    echo ">>> Check Longhorn volumes..."
+
+    volumes=$(kubectl get volumes.longhorn.io -A -o yaml)
+
+    # all volumes should be healthy
+    degraded=$(echo "$volumes" | yq '.items | any_c(.status.state == "attached" and .status.robustness != "healthy")')
```

Comment:
Do we want to skip the check when running the script against a single-node cluster?

Reply:
Good suggestion, I added checks in https://github.com/harvester/upgrade-helpers/pull/3/commits/b074f0226fa296e954f8f064d327b97bf6d11a68
And I also fix the problem that inside a piped while scope, we can't set the global variable.